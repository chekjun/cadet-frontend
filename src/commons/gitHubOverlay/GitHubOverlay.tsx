import { DialogStep, IButtonProps, ITreeNode, MultistepDialog } from '@blueprintjs/core';
import React from 'react';

import { store } from '../../pages/createStore';
import { actions } from '../utils/ActionsHelper';
import { FileExplorerPanel } from './FileExplorerPanel';
import { RepositoryExplorerPanel } from './RepositoryExplorerPanel';

export interface GitHubOverlayProps {
  // userRepos: [];
  // pickerType: string;
  isPickerOpen: boolean;
  handleEditorValueChange: (val: string) => void;
}

export interface GitHubOverlayState {
  repoName: string;
  repoFiles: ITreeNode[];
  fileIndex: number;
  filePath: string;
}

export class GitHubOverlay extends React.PureComponent<GitHubOverlayProps, GitHubOverlayState> {
  constructor(props: GitHubOverlayProps | Readonly<GitHubOverlayProps>) {
    super(props);
    this.setRepoName = this.setRepoName.bind(this);
    this.setFilePath = this.setFilePath.bind(this);
    this.createNode = this.createNode.bind(this);
    this.setRepoFiles = this.setRepoFiles.bind(this);
    this.getFileContents = this.getFileContents.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  public state: GitHubOverlayState = {
    repoName: '',
    repoFiles: [],
    fileIndex: 0,
    filePath: ''
  };

  userRepos = store.getState().session.userRepos;
  pickerType = store.getState().session.pickerType;
  isPickerOpen = store.getState().session.isPickerOpen;

  setRepoName(e: any) {
    this.setState({ repoName: e.target.value });
  }

  setFilePath(e: string) {
    this.setState({ filePath: e });
  }

  async createNode(thisFile: any) {
    const octokit = store.getState().session.githubOctokitInstance;
    const gitHubLogin = store.getState().session.gitHubLogin;
    if (octokit !== undefined) {
      const index = this.state.fileIndex++;
      if (thisFile.type === 'file') {
        const node: ITreeNode = {
          id: index,
          nodeData: thisFile.path,
          icon: 'document',
          label: thisFile.name
        };
        return node;
      }
      if (thisFile.type === 'dir') {
        const folder: ITreeNode[] = [];
        const results = await octokit.repos.getContent({
          owner: gitHubLogin,
          repo: this.state.repoName,
          path: thisFile.path
        });
        const files = results.data;
        console.log(files);
        if (Array.isArray(files)) {
          for (let i = 0; i < files.length; i++) {
            folder.push(await this.createNode(files[i]));
          }
        }
        const node: ITreeNode = {
          id: index,
          nodeData: thisFile.path,
          icon: 'folder-close',
          label: thisFile.name,
          childNodes: folder
        };
        return node;
      }
    }
    const node: ITreeNode = {
      id: this.state.fileIndex++,
      label: 'dummy file'
    };
    return node;
  }

  async setRepoFiles() {
    this.setState({ fileIndex: 0 });
    this.setState({ repoFiles: [] });
    const octokit = store.getState().session.githubOctokitInstance;
    const gitHubLogin = store.getState().session.gitHubLogin;
    if (octokit === undefined || this.state.repoName === '') return;
    try {
      const newRepoFiles: ITreeNode[] = [];
      const results = await octokit.repos.getContent({
        owner: gitHubLogin,
        repo: this.state.repoName,
        path: ''
      });
      const files = results.data;
      if (Array.isArray(files)) {
        for (let i = 0; i < files.length; i++) {
          newRepoFiles.push(await this.createNode(files[i]));
        }
      }
      this.setState({ repoFiles: newRepoFiles });
    } catch (err) {
      console.error(err);
    }
  }

  async getFileContents() {
    const octokit = store.getState().session.githubOctokitInstance;
    const gitHubLogin = store.getState().session.gitHubLogin;
    if (octokit === undefined) return;
    octokit.repos
      .getContent({
        owner: gitHubLogin,
        repo: this.state.repoName,
        path: this.state.filePath
      })
      .then(
        value => {
          // fulfillment
          const { content } = { ...value.data };
          if (content) {
            this.props.handleEditorValueChange(Buffer.from(content, 'base64').toString());
          }
        },
        reason => {
          // rejection
          console.error(reason);
        }
      );
  }

  public render() {
    const finalButtonProps: Partial<IButtonProps> = {
      intent: 'primary',
      onClick: this.handleSubmit,
      text: store.getState().session.pickerType 
    };

    return (
      <MultistepDialog
        className="GitHubPicker"
        finalButtonProps={finalButtonProps}
        isOpen={this.props.isPickerOpen}
        nextButtonProps={{ disabled: this.state.repoFiles.length === 0 }}
        onClose={this.handleClose}
        title="Opening Repository File"
      >
        <DialogStep
          id="Repository"
          panel={
            <RepositoryExplorerPanel
              userRepos={this.userRepos}
              repoName={this.state.repoName}
              setRepoName={this.setRepoName}
              setRepoFiles={this.setRepoFiles}
              {...this.props}
            />
          }
          title="Select Repository"
        />

        <DialogStep
          id="Files"
          panel={
            <FileExplorerPanel repoFiles={this.state.repoFiles} setFilePath={this.setFilePath} />
          }
          title="Select File"
        />
      </MultistepDialog>
    );
  }

  handleClose() {
    store.dispatch(actions.setPickerDialog(false));
  }

  handleSubmit() {
    this.getFileContents();
    store.dispatch(actions.setPickerDialog(false));
  }
}

/*
  async setFileContents() {
    if (this.octokit === undefined) return;
    this.octokit.repos.createOrUpdateFileContents({
      owner: this.state.gitHubLogin,
      repo: this.state.repoName,
      path: this.state.filePath,
      message: 'TEST PUSH FROM OCTOKIT',
      content: 'TEST CONTENT',
      name: this.state.gitHubName,
      email: this.state.gitHubEmail
    });
  }
*/

/*
  sortNodeList (nodeList: ITreeNode[]) {
    console.log('sorted nodelist');
    nodeList.sort(function(x, y) {
      if (x.hasCaret === false && y.hasCaret === true) {
        const tempNode = x;
        x = y;
        y = tempNode;
      } else if (x.hasCaret === y.hasCaret) {
        if (x.label > y.label) {
          const tempNode = x;
          x = y;
          y = tempNode;
        }
      }
      return -1;
    });
    nodeList.forEach(node => {
      if (node.childNodes !== undefined) {
        this.sortNodeList(node.childNodes);
      }
    });
  }
*/
