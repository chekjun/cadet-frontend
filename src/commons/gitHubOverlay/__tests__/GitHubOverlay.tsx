import { act, fireEvent, render, screen } from '@testing-library/react';
import { mount } from 'enzyme';

import GitHubOverlay from '../GitHubOverlay';

test('Test clicking close closes picker', () => {
  act(() => {
    render(<GitHubOverlay {...props} />);
  });

  const leftClick = { button: 1 };
  fireEvent.click(screen.getByLabelText('Close'), leftClick);
  expect(props.handleGitHubCloseFileExplorerDialog).toHaveBeenCalledTimes(1);
});

test('Test GitHubOverlay renders correctly', () => {
  const REP = mount(<GitHubOverlay {...props} />);
  expect(REP.debug()).toMatchSnapshot();
});

const repo1 = { name: 'repoName1', id: 1 };
const repo2 = { name: 'repoName2', id: 2 };
const repo3 = { name: 'repoName3', id: 3 };
const userRepos: { name: string; id: number }[] = [repo1, repo2, repo3];
const props = {
  userRepos: userRepos,
  pickerType: 'Open',
  isPickerOpen: true,
  handleEditorValueChange: jest.fn(),
  handleGitHubCloseFileExplorerDialog: jest.fn(),
  handleGitHubBeginConfirmationDialog: jest.fn()
};
