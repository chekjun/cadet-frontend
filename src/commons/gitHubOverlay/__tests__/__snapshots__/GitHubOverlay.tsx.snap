// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GitHubOverlay renders correctly 1`] = `
"<GitHubOverlay userRepos={{...}} pickerType=\\"Open\\" isPickerOpen={true} handleEditorValueChange={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} handleGitHubCloseFileExplorerDialog={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} handleGitHubBeginConfirmationDialog={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
  <div>
    <Blueprint3.MultistepDialog finalButtonProps={{...}} isOpen={true} onClose={[Function: bound handleClose]} title=\\"Open file\\" usePortal={false} canOutsideClickClose={true}>
      <Blueprint3.Dialog finalButtonProps={{...}} isOpen={true} onClose={[Function: bound handleClose]} title=\\"Open file\\" usePortal={false} canOutsideClickClose={true} style={{...}}>
        <Blueprint3.Overlay finalButtonProps={{...}} isOpen={true} onClose={[Function: bound handleClose]} title=\\"Open file\\" usePortal={false} canOutsideClickClose={true} style={{...}} className=\\"bp3-overlay-scroll-container\\" hasBackdrop={true} autoFocus={true} backdropProps={{...}} canEscapeKeyClose={true} enforceFocus={true} lazy={true} transitionDuration={300} transitionName=\\"bp3-overlay\\">
          <TransitionGroup appear={true} className=\\"bp3-overlay bp3-overlay-open bp3-overlay-inline bp3-overlay-scroll-container\\" component=\\"div\\" onKeyDown={[Function (anonymous)]} childFactory={[Function: childFactory]}>
            <div className=\\"bp3-overlay bp3-overlay-open bp3-overlay-inline bp3-overlay-scroll-container\\" onKeyDown={[Function (anonymous)]}>
              <CSSTransition classNames=\\"bp3-overlay\\" timeout={300} addEndListener={[Function (anonymous)]} onExited={[Function: bound bound handleExited]} in={true} appear={true} enter={[undefined]} exit={[undefined]}>
                <Transition timeout={300} addEndListener={[Function (anonymous)]} onExited={[Function (anonymous)]} in={true} appear={true} enter={true} exit={true} onEnter={[Function (anonymous)]} onEntered={[Function (anonymous)]} onEntering={[Function (anonymous)]} onExit={[Function (anonymous)]} onExiting={[Function (anonymous)]} mountOnEnter={false} unmountOnExit={false}>
                  <div className=\\"bp3-overlay-backdrop\\" onMouseDown={[Function (anonymous)]} tabIndex={0} />
                </Transition>
              </CSSTransition>
              <CSSTransition classNames=\\"bp3-overlay\\" onEntering={[undefined]} onEntered={[undefined]} onExiting={[undefined]} onExited={[Function: bound bound handleExited]} timeout={300} addEndListener={[Function (anonymous)]} in={true} appear={true} enter={[undefined]} exit={[undefined]}>
                <Transition onEntering={[Function (anonymous)]} onEntered={[Function (anonymous)]} onExiting={[Function (anonymous)]} onExited={[Function (anonymous)]} timeout={300} addEndListener={[Function (anonymous)]} in={true} appear={true} enter={true} exit={true} onEnter={[Function (anonymous)]} onExit={[Function (anonymous)]} mountOnEnter={false} unmountOnExit={false}>
                  <div className=\\"bp3-dialog-container bp3-overlay-content\\" tabIndex={0}>
                    <div className=\\"bp3-dialog\\" style={{...}}>
                      <div className=\\"bp3-dialog-header\\">
                        <Blueprint3.Icon icon={[undefined]} iconSize={20} />
                        <Component>
                          <h4 className=\\"bp3-heading\\">
                            Open file
                          </h4>
                        </Component>
                        <Blueprint3.Button aria-label=\\"Close\\" className=\\"bp3-dialog-close-button\\" icon={{...}} minimal={true} onClick={[Function: bound handleClose]}>
                          <button type=\\"button\\" aria-label=\\"Close\\" className=\\"bp3-button bp3-minimal bp3-dialog-close-button\\" onClick={[Function: bound handleClose]} disabled={[undefined]} onBlur={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
                            <Blueprint3.Icon icon={{...}}>
                              <Blueprint3.Icon icon=\\"small-cross\\" iconSize={20}>
                                <span icon=\\"small-cross\\" className=\\"bp3-icon bp3-icon-small-cross\\" title={[undefined]}>
                                  <svg fill={[undefined]} data-icon=\\"small-cross\\" width={20} height={20} viewBox=\\"0 0 20 20\\">
                                    <desc>
                                      small-cross
                                    </desc>
                                    <path d=\\"M11.41 10l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L10 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L8.59 10 5.3 13.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l3.29-3.3 3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L11.41 10z\\" fillRule=\\"evenodd\\" />
                                  </svg>
                                </span>
                              </Blueprint3.Icon>
                            </Blueprint3.Icon>
                            <Blueprint3.Icon icon={[undefined]} />
                          </button>
                        </Blueprint3.Button>
                      </div>
                      <div className=\\"bp3-multistep-dialog-panels\\">
                        <div className=\\"bp3-multistep-dialog-left-panel\\">
                          <div className=\\"bp3-dialog-step-container bp3-active\\">
                            <div className=\\"bp3-dialog-step\\" onClick={[Function (anonymous)]}>
                              <div className=\\"bp3-dialog-step-icon\\">
                                1
                              </div>
                              <div className=\\"bp3-dialog-step-title bp3-active\\">
                                Select Repository
                              </div>
                            </div>
                          </div>
                          <div className=\\"bp3-dialog-step-container\\">
                            <div className=\\"bp3-dialog-step\\" onClick={[undefined]}>
                              <div className=\\"bp3-dialog-step-icon\\">
                                2
                              </div>
                              <div className=\\"bp3-dialog-step-title\\">
                                Select File
                              </div>
                            </div>
                          </div>
                        </div>
                        <div className=\\"bp3-multistep-dialog-right-panel\\">
                          <RepositoryExplorerPanel userRepos={{...}} repoName=\\"\\" setRepoName={[Function: bound setRepoName]} refreshRepoFiles={[Function: bound refreshRepoFiles]} pickerType=\\"Open\\" isPickerOpen={true} handleEditorValueChange={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} handleGitHubCloseFileExplorerDialog={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} handleGitHubBeginConfirmationDialog={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
                            <div className=\\"bp3-dialog-body\\">
                              <Blueprint3.RadioGroup className=\\"RepoPanel\\" onChange={[Function: bound setRepoName]} selectedValue=\\"\\">
                                <div className=\\"RepoPanel\\">
                                  <Blueprint3.Radio label=\\"repoName1\\" value=\\"repoName1\\" checked={false} className={[undefined]} disabled={[undefined]} inline={[undefined]} name=\\"Blueprint3.RadioGroup-1\\" onChange={[Function: bound setRepoName]}>
                                    <Control label=\\"repoName1\\" value=\\"repoName1\\" checked={false} className={[undefined]} disabled={[undefined]} inline={[undefined]} name=\\"Blueprint3.RadioGroup-1\\" onChange={[Function: bound setRepoName]} type=\\"radio\\" typeClassName=\\"bp3-radio\\">
                                      <label className=\\"bp3-control bp3-radio\\" style={[undefined]}>
                                        <input value=\\"repoName1\\" checked={false} disabled={[undefined]} name=\\"Blueprint3.RadioGroup-1\\" onChange={[Function: bound setRepoName]} type=\\"radio\\" />
                                        <span className=\\"bp3-control-indicator\\" />
                                        repoName1
                                      </label>
                                    </Control>
                                  </Blueprint3.Radio>
                                  <Blueprint3.Radio label=\\"repoName2\\" value=\\"repoName2\\" checked={false} className={[undefined]} disabled={[undefined]} inline={[undefined]} name=\\"Blueprint3.RadioGroup-1\\" onChange={[Function: bound setRepoName]}>
                                    <Control label=\\"repoName2\\" value=\\"repoName2\\" checked={false} className={[undefined]} disabled={[undefined]} inline={[undefined]} name=\\"Blueprint3.RadioGroup-1\\" onChange={[Function: bound setRepoName]} type=\\"radio\\" typeClassName=\\"bp3-radio\\">
                                      <label className=\\"bp3-control bp3-radio\\" style={[undefined]}>
                                        <input value=\\"repoName2\\" checked={false} disabled={[undefined]} name=\\"Blueprint3.RadioGroup-1\\" onChange={[Function: bound setRepoName]} type=\\"radio\\" />
                                        <span className=\\"bp3-control-indicator\\" />
                                        repoName2
                                      </label>
                                    </Control>
                                  </Blueprint3.Radio>
                                  <Blueprint3.Radio label=\\"repoName3\\" value=\\"repoName3\\" checked={false} className={[undefined]} disabled={[undefined]} inline={[undefined]} name=\\"Blueprint3.RadioGroup-1\\" onChange={[Function: bound setRepoName]}>
                                    <Control label=\\"repoName3\\" value=\\"repoName3\\" checked={false} className={[undefined]} disabled={[undefined]} inline={[undefined]} name=\\"Blueprint3.RadioGroup-1\\" onChange={[Function: bound setRepoName]} type=\\"radio\\" typeClassName=\\"bp3-radio\\">
                                      <label className=\\"bp3-control bp3-radio\\" style={[undefined]}>
                                        <input value=\\"repoName3\\" checked={false} disabled={[undefined]} name=\\"Blueprint3.RadioGroup-1\\" onChange={[Function: bound setRepoName]} type=\\"radio\\" />
                                        <span className=\\"bp3-control-indicator\\" />
                                        repoName3
                                      </label>
                                    </Control>
                                  </Blueprint3.Radio>
                                </div>
                              </Blueprint3.RadioGroup>
                            </div>
                          </RepositoryExplorerPanel>
                          <div className=\\"bp3-multistep-dialog-footer\\">
                            <div className=\\"bp3-dialog-footer-actions\\">
                              <Blueprint3.Button intent=\\"primary\\" onClick={[Function (anonymous)]} text=\\"Next\\">
                                <button type=\\"button\\" onClick={[Function (anonymous)]} className=\\"bp3-button bp3-intent-primary\\" disabled={[undefined]} onBlur={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
                                  <Blueprint3.Icon icon={[undefined]} />
                                  <span className=\\"bp3-button-text\\">
                                    Next
                                  </span>
                                  <Blueprint3.Icon icon={[undefined]} />
                                </button>
                              </Blueprint3.Button>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </Transition>
              </CSSTransition>
            </div>
          </TransitionGroup>
        </Blueprint3.Overlay>
      </Blueprint3.Dialog>
    </Blueprint3.MultistepDialog>
  </div>
</GitHubOverlay>"
`;
